File: agents/addon/extension/openai_chatgpt/openai_chatgpt.go
/**
 *
 * Agora Real Time Engagement
 * Created by lixinhui in 2024.
 * Copyright (c) 2024 Agora IO. All rights reserved.
 *
 */
// Note that this is just an example extension written in the GO programming
// language, so the package name does not equal to the containing directory
// name. However, it is not common in Go.
package extension

import (
	"context"
	"fmt"
	"math/rand"
	"net/http"
	"net/url"

	openai "github.com/sashabaranov/go-openai"
)

type openaiChatGPT struct {
	client *openai.Client
	config openaiChatGPTConfig
}

type openaiChatGPTConfig struct {
	ApiKey string

	Model  string
	Prompt string

	FrequencyPenalty float32
	PresencePenalty  float32
	TopP             float32
	Temperature      float32
	MaxTokens        int
	Seed             int

	ProxyUrl string
}

func defaultOpenaiChatGPTConfig() openaiChatGPTConfig {
	return openaiChatGPTConfig{
		ApiKey: "",

		Model:  openai.GPT4o,
		Prompt: "You are a voice assistant who talks in a conversational way and can chat with me like my friends. i will speak to you in english or chinese, and you will answer in the corrected and improved version of my text with the language i use. Don't talk like a robot, instead i would like you to talk like real human with emotions. i will use your answer for text-to-speech, so don't return me any meaningless characters. I want you to be helpful, when i'm asking you for advices, give me precise, practical and useful advices instead of being vague. When giving me list of options, express the options in a narrative way instead of bullet points.",

		FrequencyPenalty: 0.9,
		PresencePenalty:  0.9,
		TopP:             1.0,
		Temperature:      0.1,
		MaxTokens:        512,
		Seed:             rand.Int(),

		ProxyUrl: "",
	}
}

func newOpenaiChatGPT(config openaiChatGPTConfig) (*openaiChatGPT, error) {
	conf := openai.DefaultConfig(config.ApiKey)
	if config.ProxyUrl != "" {
		proxyUrl, err := url.Parse(config.ProxyUrl)
		if err != nil {
			return nil, fmt.Errorf("newOpenaiChatGPT failed on parsing proxy url, err: %v", err)
		}
		conf.HTTPClient = &http.Client{Transport: &http.Transport{Proxy: http.ProxyURL(proxyUrl)}}
	}

	return &openaiChatGPT{
		config: config,
		client: openai.NewClientWithConfig(conf),
	}, nil
}

func (c *openaiChatGPT) getChatCompletionsStream(messages []openai.ChatCompletionMessage) (*openai.ChatCompletionStream, error) {
	req := openai.ChatCompletionRequest{
		Temperature:      c.config.Temperature,
		TopP:             c.config.TopP,
		PresencePenalty:  c.config.PresencePenalty,
		FrequencyPenalty: c.config.FrequencyPenalty,
		MaxTokens:        c.config.MaxTokens,
		Seed:             &c.config.Seed,
		Messages: append(
			[]openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleSystem,
					Content: c.config.Prompt,
				},
			},
			messages...,
		),
		Model:  c.config.Model,
		Stream: true,
	}

	resp, err := c.client.CreateChatCompletionStream(context.Background(), req)
	if err != nil {
		return nil, fmt.Errorf("CreateChatCompletionStream failed,err: %v", err)
	}
	return resp, nil
}

File: agents/addon/extension/openai_chatgpt_python/openai_chatgpt.py
import random
import requests
import openai
from typing import List, Dict, Any, Optional


class OpenAIChatGPTConfig:
    def __init__(self, 
            base_url: str, 
            api_key: str, 
            model: str, 
            prompt: str, 
            frequency_penalty: float, 
            presence_penalty: float, 
            top_p: float, 
            temperature: float, 
            max_tokens: int, 
            seed: Optional[int] = None, 
            proxy_url: Optional[str] = None):
        self.base_url = base_url
        self.api_key = api_key
        self.model = model
        self.prompt = prompt
        self.frequency_penalty = frequency_penalty
        self.presence_penalty = presence_penalty
        self.top_p = top_p
        self.temperature = temperature
        self.max_tokens = max_tokens
        self.seed = seed if seed is not None else random.randint(0, 10000)
        self.proxy_url = proxy_url

    @classmethod
    def default_config(cls):
        return cls(
            base_url="https://api.openai.com/v1",
            api_key="",
            model="gpt-4",  # Adjust this to match the equivalent of `openai.GPT4o` in the Python library
            prompt="You are a voice assistant who talks in a conversational way and can chat with me like my friends. I will speak to you in English or Chinese, and you will answer in the corrected and improved version of my text with the language I use. Don’t talk like a robot, instead I would like you to talk like a real human with emotions. I will use your answer for text-to-speech, so don’t return me any meaningless characters. I want you to be helpful, when I’m asking you for advice, give me precise, practical and useful advice instead of being vague. When giving me a list of options, express the options in a narrative way instead of bullet points.",
            frequency_penalty=0.9,
            presence_penalty=0.9,
            top_p=1.0,
            temperature=0.1,
            max_tokens=512,
            seed=random.randint(0, 10000),
            proxy_url=""
        )
    

class OpenAIChatGPT:
    def __init__(self, config: OpenAIChatGPTConfig):
        self.config = config
        openai.api_key = config.api_key
        openai.api_base = config.base_url
        if config.proxy_url:
            proxies = {
                "http": config.proxy_url,
                "https": config.proxy_url,
            }
            self.session = requests.Session()
            self.session.proxies.update(proxies)
        else:
            self.session = requests.Session()

    def get_chat_completions_stream(self, messages):
        req = {
            "model": self.config["Model"],
            "messages": [
                {
                    "role": "system",
                    "content": self.config["Prompt"],
                },
                *messages,
            ],
            "temperature": self.config["Temperature"],
            "top_p": self.config["TopP"],
            "presence_penalty": self.config["PresencePenalty"],
            "frequency_penalty": self.config["FrequencyPenalty"],
            "max_tokens": self.config["MaxTokens"],
            "seed": self.config["Seed"],
            "stream": True,
        }

        try:
            response = self.client.Completion.create(**req)
            return response
        except Exception as e:
            raise Exception(f"CreateChatCompletionStream failed, err: {e}")